distributed/tests/test_worker.py

```diff
@@ -3161,85 +3161,84 @@ async def get_data(self, comm, *args, **kwargs):
         return await super().get_data(comm, *args, **kwargs)
 
 
-@gen_cluster(client=True, nthreads=[])
-async def test_gather_dep_cancelled_rescheduled(c, s):
-    """At time of writing, the gather_dep implementation filtered tasks again
-    for in-flight state. The response parser, however, did not distinguish
-    resulting in unwanted missing-data signals to the scheduler, causing
-    potential rescheduling or data leaks.
-
-    If a cancelled key is rescheduled for fetching while gather_dep waits
-    internally for get_data, the response parser would misclassify this key and
-    cause the key to be recommended for a release causing deadlocks and/or lost
-    keys.
-    At time of writing, this transition was implemented wrongly and caused a
-    flight->cancelled transition which should be recoverable but the cancelled
-    state was corrupted by this transition since ts.done==True. This attribute
-    setting would cause a cancelled->fetch transition to actually drop the key
-    instead, causing https://github.com/dask/distributed/issues/5366
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_gather_dep_cancelled_rescheduled(c, s, a):
+    """A task transitions flight->cancelled->fetch->flight, all while gather_dep is
+    waiting for the data of the initial flight.
```

Comment:
This test can now be drastically simplified after #6371.
Remove assumption that both dependencies of a task will be gathered in a single fetch (this will no longer be the case in #6462).

Reply:
> This test can now be drastically simplified after #6371.

I don't think we should simplify tests just because we know that internals changed.

> Remove assumption that both dependencies of a task will be gathered in a single fetch

Where do we assume this in the previous test?

Reply:
> I don't think we should simplify tests just because we know that internals changed.

The previous test was so complicated because it had to have two checkpoints:

1. after you enter gather_dep, but before you run the preamble
2. while you're in the comms

I think that being able to say "it doesn't really matter in which point of gather_dep you pinch it; nothing fancy will happen before comms" is a very reasonable thing to say?

> Where do we assume this in the previous test?
```python
fut4 = c.submit(sum, fut1, fut2, workers=[b.address], key="f4")
```
This sends a compute-task event to b, which in turn sends f1 and f2 into fetch at the same time.
As they're dependencies of the same task, they have the same exact priority.

After #6462, there are two separate fetches. You'll get that f1 is fetched first 50% of the times, at random. The two keys are most likely go through a set at some point during the process, and the ordering of sets changes every time you restart the python interpreter.

Reply:
> I think that being able to say "it doesn't really matter in which point of gather_dep you pinch it; nothing fancy will happen before comms" is a very reasonable thing to say?

Will this be true forever? How do we ensure that this stays true? This test asks the question 
"Is it harmful to cancel a task after a fetch is scheduled (`gather-dep`) but before we're actually stepping in this coroutine".
The only way to answer this question with certainty is to have this test or ensure there is not a single await before the get_data message is sent which is hard/impossible.

While this is a seemingly artificial question to ask, this was causing a deadlock in the past https://github.com/dask/distributed/pull/5525 and I don't think the pre-filtering optimization was entirely unreasonable.

Test are also there to protect us from reintroducing regressions and this test is written at a high enough abstraction level (we only rely on task state names (e.g. flight, cancelled) and the fact that there are the gather_dep/get_data methods) that I have confidence it won't bother us a lot. The test is not known to be flaky, it runs very fast (0.1s) and still works on your proposed change of https://github.com/dask/distributed/pull/6462 (at least for me locally)

I don't see a reason why we should change anything about this test